### PR TITLE
Multi actor component

### PR DESCRIPTION
**Pull request by laughxing**

### CODE REVIEW

File: Source/ChanneldUE/ChanneldNetDriver.cpp [hunk 1]
```diff
@@ -281,7 +281,18 @@ void UChanneldNetDriver::HandleCustomRPC(TSharedPtr<unrealpb::RemoteFunctionMess
...
+	UObject* SubObject = nullptr;
+	if (Msg->subobjectpath().length() > 0)
+	{
+		FString ComponentClassPath = UTF8_TO_TCHAR(Msg->subobjectpath().c_str());
+		UClass* ComponentClass = FindObject<UClass>(ANY_PACKAGE, *ComponentClassPath);
+		if (ComponentClass == nullptr)
+		{
+			ComponentClass = LoadClass<UActorComponent>(NULL,  *ComponentClassPath);
+		}
+		SubObject = Actor->GetComponentByClass(ComponentClass);
```
Comment (indiest) [inline Source/ChanneldUE/ChanneldNetDriver.cpp:293]: Is it possible that we use `GetDefaultSubobjectByName` instead of `GetComponentByClass`  so
1) We can support multiple ActorComponents of the same class (e.g. SceneComponent);
2) The RPC payload is optimized as the component name is generally shorter than the class path.

File: Source/ChanneldUE/ChanneldUtils.cpp [hunk 2]
```diff
@@ -347,12 +347,15 @@ UActorComponent* ChanneldUtils::GetActorComponentByRefChecked(const unrealpb::Ac
 	}
 
 	FName CompName = FName(UTF8_TO_TCHAR(Ref->compname().c_str()));
-	UObject* Comp = Actor->GetDefaultSubobjectByName(CompName);
-	if (Comp)
+
+	auto Components = Actor->GetComponents();
+	for(auto C : Components)
```
Comment (indiest) [inline Source/ChanneldUE/ChanneldUtils.cpp:352]: Is `GetDefaultSubobjectByName` not working here?

File: Source/ChanneldUE/channeld.pb.h [hunk 3]
```diff
@@ -287,6 +287,7 @@ enum ChannelType : int {
   SUBWORLD = 3,
   SPATIAL = 4,
   ENTITY = 5,
+  PARALLEL_SERVER = 6,
```
Comment (indiest) [inline Source/ChanneldUE/channeld.pb.h:290]: Maybe this change should be in another PR?

File: Source/ChanneldUE/unreal_common.pb.h
Comment (indiest) [inline Source/ChanneldUE/unreal_common.pb.h:1]: What are the changes in `unreal_common.proto`?

File: Source/ChanneldUE/Replication/ChanneldActorComponentReplicator.cpp [hunk 4]
```diff
@@ -9,6 +9,7 @@ FChanneldActorComponentReplicator::FChanneldActorComponentReplicator(UObject* In
 	TArray<FLifetimeProperty> RepProps;
 	DisableAllReplicatedPropertiesOfClass(InTargetObj->GetClass(), UActorComponent::StaticClass(), EFieldIteratorFlags::ExcludeSuper, RepProps);
 
+	TargetComponentName = TCHAR_TO_UTF8(*InTargetObj->GetName());
```
Comment (indiest) [inline Source/ChanneldUE/Replication/ChanneldActorComponentReplicator.cpp:12]: Is `TargetComponentName` field used elsewhere in the cpp?

File: Source/ChanneldUE/unreal_common.pb.h
Comment (indiest) [reply Source/ChanneldUE/unreal_common.pb.h:1]: Nevermind. Found https://github.com/metaworking/channeld/pull/44

File: Source/ChanneldUE/channeld.pb.h (hunk 3 above)
Comment (laughxing) [reply Source/ChanneldUE/channeld.pb.h:290]: this feature is not ready for a PR, I will remove this

File: Source/ChanneldUE/ChanneldUtils.cpp (hunk 2 above)
Comment (laughxing) [reply Source/ChanneldUE/ChanneldUtils.cpp:352]: GetDefaultSubobjectByName failed on our test game, so changed to this. but haven't had time to dive into

Comment (indiest) [reply Source/ChanneldUE/ChanneldUtils.cpp:352]: Could you please send the code that reproduce the failure of GetDefaultSubobjectByName so I can test?

File: Source/ChanneldUE/Replication/ChanneldActorComponentReplicator.cpp (hunk 4 above)
Comment (laughxing) [reply Source/ChanneldUE/Replication/ChanneldActorComponentReplicator.cpp:12]: fixed in the new commit

File: Source/ChanneldUE/Replication/ChanneldActorComponentReplicator.cpp [hunk 5]
```diff
@@ -51,6 +52,7 @@ void FChanneldActorComponentReplicator::Tick(float DeltaTime)
 	if (bStateChanged)
 	{
 		FullState->MergeFrom(*DeltaState);
+		FullState->set_compname(TargetComponentName);
```
Comment (indiest) [inline Source/ChanneldUE/Replication/ChanneldActorComponentReplicator.cpp:55]: Did you mean `DeltaState->set_compname(TargetComponentName);` (if you want to send `TargetComponentName` to channeld) ?

Comment (laughxing) [reply Source/ChanneldUE/Replication/ChanneldActorComponentReplicator.cpp:55]: this is not necessary, I will check the code later

**Comment by indiest** [comment]
Hi @laughxing, the review is almost done, except my feedback on the `GetDefaultSubobjectByName`. I think the validation of this function is vital for this feature, and could be important for the future development of ChanneldUE.

**Comment by laughxing** [comment]
@indiest I happened to have a free morning to test this. but have not find a way to reproduce the problem on [channeld-ue-demos](https://github.com/metaworking/channeld-ue-demos) project.
this is the call stack on our project using ue5.1.
image
in case our code doesn't match, here's the code around line 146  of ChanneldCharacterReplicator.cpp
image

with the warning log

>  Warning: GetActorComponentByRefChecked: Cannot find component 'LandscapeHeightfieldCollisionComponent_426' of actor LandscapeStreamingProxy_D7T4VF4LBP34PHV9XOEVM5TIG_1_5_5_0

It seems that this problem is related to the World Partition feature, but I am unable to reproduce the warning on the channeld-ue-demo project

**Comment by indiest** [comment]
I see.

What is the World Partition setup of your project? 

Do you use it together with the spatial channels? 
If so, did you call `GetDefaultSubobjectByName` on `LandscapeStreamingProxy_D7T4VF4LBP34PHV9XOEVM5TIG_1_5_5_0` in a different server?

**Comment by laughxing** [comment]
I'm testing on SinglechannelDataView

**Comment by indiest** [comment]
Reproduced the issue:
>Warning: GetActorComponentByRefChecked: Cannot find component 'LandscapeHeightfieldCollisionComponent_426' of actor LandscapeStreamingProxy_D7T4VF4LBP34PHV9XOEVM5TIG_1_5_5_0

The cause of the issue is that `GetDefaultSubobjectByName` can only retrive the static/default subobject of an UObject, but the collision components of the Landscape are generate dynamically.

Fixed the issue in this commit: 01f407d3f7ff0b182a41211dd7b5ab85e6f64bbc

**Comment by laughxing** [comment]
@indiest the unreal_common.pb.h/cpp file is using version from master,  need to generate a new version with https://github.com/metaworking/channeld/pull/44 merged to make it compile

**Comment by indiest** [comment]
No problem. I will merge the PR in channeld and re-generate the protobuf code.
